### PR TITLE
[5.2] Allow models to specify a timezone for $dates

### PR DESCRIPTION
**Pull request by natesilva**
Date mutators depend on a global variable, PHP's default timezone. This can cause incorrect timestamps to be saved to and retrieved from the database if the timezone ever changes, or if two environments use a different timezone setting:

``` php
// assume: global timezone is UTC
// assume: my_timestamp column is listed in $dates or $casts

$x = App\SomeModel::find(1);
$y = $x->my_timestamp;              // perhaps 2016-01-02 03:04:05 UTC

// timezone changed (or it's a different app, or the app is running
// with different environment settings, etc.)

date_default_timezone_set('America/New_York');
$z = $x->my_timestamp;
// $z is 2016-01-02 03:04:05 EST
// this is NOT equivalent to the above UTC timestamp
```

This patch adds a `Model` property, `$timezone`. The default value is `null` and behaves exacty as it does today.

If set to a non-null value, timestamp columns listed in `$dates`, or listed in `$casts` as `"datetime"`, will be rendered in the given timezone when written to the database and to the equivalent local time when retrieved from the database:

``` php
// assume: global timezone is UTC
// assume: my_timestamp column is listed in $dates or $casts
// assume: the model's $timestamp value is set to 'UTC'

$x = App\SomeModel::find(1);
$y = $x->my_timestamp;              // perhaps 2016-01-02 03:04:05 UTC

// timezone changed
date_default_timezone_set('America/New_York');
$z = $x->my_timestamp;
// $z is 2016-01-01 22:04:05 EST
// this IS equivalent to the above UTC timestamp
```
### Relationship to Laravel's app.timezone value

It's common to set the timezone in `config/app.php` and it's unlikely that you would change it elsewhere.

However, this is not always the case when `Eloquent` is used as a standalone ORM, for example, in a legacy non-Laravel app.

The TL;DR is that timestamp values saved to the database should be deterministic, regardless of the global timezone setting or changes to the local environment.

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Model.php [hunk 1]
```diff
@@ -141,6 +141,13 @@
     protected $dates = [];
 
     /**
+     * The timezone to use when mutating dates.
+     *
+     * @var string
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: string|null

**Comment by JosephSilber** [comment]
Shouldn't this be a static property?

**Comment by natesilva** [comment]
@JosephSilber Yes and no. Yes: it applies to all instances of the model, so should be static. No: Other similar properties in Eloquent are protected: `$table`, `$dates`, `$casts`, etc. For consistency I defined it as protected. Hopefully a future release of Eloquent will use static properties.

**Comment by JosephSilber** [comment]
Huh??

All those other properties you mentioned are model-specific properties.

The timezone is a property that should be shared between all models.

And if it's a static property, Laravel core can set it to the config timezone from the database service provider.

**Comment by natesilva** [comment]
Okay, that makes sense. I implemented it per-model-class but it seems unlikely you’d need it to be different from one model to the next.

**Comment by srmklive** [comment]
I totally agree with @JosephSilber on this. Timezone is a property that should be shared between all models. If it is set as a static property, then Laravel can directly load from the timezone property in config.

**Comment by taylorotwell** [comment]
If we make it shared across all models we're just asking for a later PR asking it to be model specific.

File: src/Illuminate/Database/Eloquent/Model.php [hunk 2]
```diff
@@ -2955,7 +2965,12 @@ protected function asDateTime($value)
         // Finally, we will just assume this date is in the format used by default on
         // the database connection and use that format to create the Carbon object
         // that is returned back out to the developers after we convert it here.
-        return Carbon::createFromFormat($this->getDateFormat(), $value);
+        if (static::$timezone) {
+            return Carbon::createFromFormat($this->getDateFormat(), $value, static::$timezone)
+                ->setTimezone(date_default_timezone_get());
```
Comment (taylorotwell) [inline src/Illuminate/Database/Eloquent/Model.php]: Why do you get the global PHP timezone here? I thought the point of this was to avoid that?

**Comment by JosephSilber** [comment]
@taylorotwell 

1. I don't see how this makes sense to be model specific.
2. If it's model specific, this can't be handled by Laravel automatically. Big drawback.

File: src/Illuminate/Database/Eloquent/Model.php (hunk 2 above)
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: Yeh, this doesn't look correct.

Comment (natesilva) [inline src/Illuminate/Database/Eloquent/Model.php]: This takes the date from the database, which is stored in (for example) UTC format, and creates a `DateTimeInterface` or `Carbon` object that is in the PHP default timezone. In a Laravel app your PHP default timezone is `Config::get('app.timezone')`.

In other words, you've specified what timezone the database is using, but your app may not be using that timezone. You would expect the ORM to handle this detail for you, returning a DateTime object in the timezone of your app. This DateTime object is guaranteed to be equivalent to the date/time stored in the database, no matter what timezone your app is using. And of course, anything with `DateTimeInterface` can losslessly be converted to another timezone if desired.

Comment (natesilva) [inline src/Illuminate/Database/Eloquent/Model.php]: The date as stored and retrieved from the database is NOT affected by the PHP default timezone. We are just converting the date, after being retrieved, into the PHP default timezone, which is probably what the programmer expects.

**Comment by taylorotwell** [comment]
Because someone will want to make it model specific.

**Comment by taylorotwell** [comment]
For now just override fromDateTime and asDateTime in a base model and you can solve this in user land.
